pkg/healthautoexport/types.go

```diff
@@ -25,12 +28,26 @@ type PayloadData struct {
 
 // Metric defines a single measurement with units, as well as time-series data points.
 type Metric struct {
-	Name  string       `json:"name"`
-	Units Units        `json:"units"`
-	Data  []*Datapoint `json:"data"`
+	Name          string           `json:"name"`
+	Units         Units            `json:"units"`
+	Data          []*Datapoint     `json:"data"`
+	SleepAnalyses []*SleepAnalysis `json:"-"`
```

Comment:
Hmm.. Seeing that we need to actually use `json.RawMessage` to handle different types for `Metric` now, I'm more inclined to actually make `Metric` like as follows:

```go
type Metric struct {
	Name  string       `json:"name"`
	Units Units        `json:"units"`

	// Raw JSON data.
	Data  json.RawMessage `json:"data"`

	// Additional fields that will be unmarshaled from Data.
	Datapoints     []*Datapoint `json:"-"`
	SleepAnalyses  []*SleepAnalysis `json:"-"`
}
```

I think this also makes it more similar to the "style" adopted by the `Workout` type. What do you think?

Reply:
I like this because it removes the anonymous type in *Metric.UnmarshalJSON(), but I'm not quite sure how the fixtures and the marshal tests should be run if we make this change. I'm seeing this as we remove the *Metric.MarshalJSON method and just let it marshal the Data field normally. What would the `PayloadMetricsSleepAnalysis` fixture look like in this case? Just pre-encoded json in the Data field? That's not very much of a test... 

Let me know what you think, I think this is the only thing currently blocking reimplementing in this style.

Reply:
Actually my thoughts were as follows:

* `(*Metric).UnmarshalJSON` will be responsible for unmarshaling data into `Data` first, and then do a second pass to unmarshal from `Data` into `SleepAnalyses` based on some conditions, or fall back to unmarshaling to `Datapoints`.
* `(*Metric).MarshalJSON` will be responsible for writing the fields from `SleepAnalyses` or `Datapoints` back into `Data`, then performing a final marshal into a `[]byte`.

But now that I think of it, this doesn't necessarily make sense, since the `Data` field never needs to be used outside before/after marshaling/unmarshaling. I think a separate type may make the most sense here just for marshaling/unmarshaling:

```go
// Metric defines a single measurement with units, as well as time-series data points.
type Metric struct {
	Name  string `json:"name"`
	Units Units  `json:"units"`

	// These fields will be marshaled/unmarshaled using the custom MarshalJSON/UnmarshalJSON implementation.
	Datapoints    []*Datapoint     `json:"-"`
	SleepAnalyses []*SleepAnalysis `json:"-"`
}

// internalMetric is an intermediate type used for JSON marshaling/unmarshaling.
type internalMetric struct {
	*metricCopy
	Data json.RawMessage `json:"data"`
}

func (m *Metric) UnmarshalJSON(bytes []byte) error {
	// Pass reference of current object into the internalMetric to be unmarshaled automatically.
	intermediate := internalMetric{
		metricCopy: (*metricCopy)(m),
	}
	if err := jsoniter.Unmarshal(bytes, &intermediate); err != nil {
		return err
	}

	// Handle unmarshaling of raw data here...
}
```

This way, we don't need to overload the `json:"data"` struct tag onto `Datapoints`, and we have to explicitly handle marshaling of both `SleepAnalyses` and `Datapoints` (using the same logic used in UnmarshalJSON). We can also reuse most of your implementation here :)

To answer your original question, I think this means that the MarshalJSON test remains the same regardless, as we should be able to marshal data from `SleepAnalyses` back into the `"data"` JSON field.

P.S. You can also decide whether to rename `Data` to `Datapoints` above, I only renamed it in this context in order to make a distinction between the `json.RawMessage` and `[]*Datapoint` fields.

Reply:
Style changes implemented as discussed.